src/integrant/core.cljc

```diff
@@ -8,10 +8,54 @@
             [weavejester.dependency :as dep]))
 
 (defprotocol RefLike
-  (ref-key [r] "Return the key of the reference."))
+  (ref-key [r] "Return the key of the reference.")
+  (ref-resolve [r config resolvef] "Return the resolved value."))
 
-(defrecord Ref    [key] RefLike (ref-key [_] key))
-(defrecord RefSet [key] RefLike (ref-key [_] key))
+(defonce
+  ^{:doc "Return a unique keyword that is derived from an ordered collection of
+  keywords. The function will return the same keyword for the same collection."
+    :arglists '([kws])}
+  composite-keyword
+  (memoize
+    (fn [kws]
+      (let [parts     (for [kw kws] (str (namespace kw) "." (name kw)))
+            prefix    (str (str/join "+" parts) "_")
+            composite (keyword "integrant.composite" (str (gensym prefix)))]
+        (doseq [kw kws] (derive composite kw))
+        composite))))
+
+(defn- normalize-key [k]
+  (if (vector? k) (composite-keyword k) k))
+
+(defn derived-from?
+  "Return true if a key is derived from candidate keyword or vector of
+  keywords."
+  [key candidate]
+  (let [key (normalize-key key)]
+    (if (vector? candidate)
+      (every? #(isa? key %) candidate)
+      (isa? key candidate))))
+
+(defn find-derived
+  "Return a seq of all entries in a map, m, where the key is derived from the
+  a candidate key, k. If there are no matching keys, nil is returned. The
+  candidate key may be a keyword, or vector of keywords."
+  [m k]
+  (seq (filter #(or (= (key %) k) (derived-from? (key %) k)) m)))
+
+(defrecord Ref [key]
+  RefLike
+  (ref-key [_] key)
+  (ref-resolve [ref config resolvef]
+    (let [[k v] (first (find-derived config (ref-key ref)))]
```

Comment:
`(refkey)` would be replaced by `key` here, and the `ref` argument changed to `_`.